<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Hae Kyung Im" />

<meta name="date" content="2021-07-12" />

<title>QGT Analysis Plan - 2022</title>

<script src="site_libs/header-attrs-2.5/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">QGT-Columbia-HKI</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hakyimlab/QGT-Columbia-HKI">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">QGT Analysis Plan - 2022</h1>
<h4 class="author">Hae Kyung Im</h4>
<h4 class="date">2021-07-12</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-07-12
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>QGT-Columbia-HKI-repo/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200603code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20200603)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200603code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200603)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomhakyimlabQGTColumbiaHKIrepotreedafeef4792509bca64ab5d684a7c13f3659cd0fftargetblankdafeef4a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/hakyimlab/QGT-Columbia-HKI-repo/tree/dafeef4792509bca64ab5d684a7c13f3659cd0ff" target="_blank">dafeef4</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomhakyimlabQGTColumbiaHKIrepotreedafeef4792509bca64ab5d684a7c13f3659cd0fftargetblankdafeef4a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/hakyimlab/QGT-Columbia-HKI-repo/tree/dafeef4792509bca64ab5d684a7c13f3659cd0ff" target="_blank">dafeef4</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
working directory clean
</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/2022_analysis_plan.Rmd</code>) and HTML (<code>docs/2022_analysis_plan.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI-repo/blob/dafeef4792509bca64ab5d684a7c13f3659cd0ff/analysis/2022_analysis_plan.Rmd" target="_blank">dafeef4</a>
</td>
<td>
Fnyasimi
</td>
<td>
2021-07-12
</td>
<td>
Update the analysis plan
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hakyimlab/QGT-Columbia-HKI-repo/e5f98925da7092faf66eac7e32cf3071bfc3ac10/docs/2022_analysis_plan.html" target="_blank">e5f9892</a>
</td>
<td>
Fnyasimi
</td>
<td>
2021-07-12
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI-repo/blob/df422b9ce11552994718c052be1ae9d3707e704d/analysis/2022_analysis_plan.Rmd" target="_blank">df422b9</a>
</td>
<td>
Fnyasimi
</td>
<td>
2021-07-12
</td>
<td>
wflow_publish(“analysis/2022_analysis_plan.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>Here we will perform some of the analyses we discussed in the lecture this morning to gain hands on experience. Follow the steps listed below to get your RStudio environment<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>.</p>
<div id="getting-access-to-the-rstudio-environment" class="section level1">
<h1>Getting access to the Rstudio environment</h1>
<ul class="task-list">
<li><input type="checkbox" disabled="" />
Find your Rstudio server IP address and get the username and password <a href="https://docs.google.com/spreadsheets/d/1W_bFGB_ctLtUB3WTgAee4PohzKR-2lOKw9kTI0RVJr8/edit?usp=sharing">here</a></li>
<li><input type="checkbox" disabled="" />
connect to the Rstudio server using the url you claimed (<a href="http://xxx.xxx.xxx.xxx:8787" class="uri">http://xxx.xxx.xxx.xxx:8787</a>) using a web browser</li>
<li><input type="checkbox" disabled="" />
log into the server using the username and password of the server you claimed</li>
</ul>
<div id="summary-of-analysis-plan" class="section level2">
<h2>Summary of analysis plan</h2>
<ul>
<li>predict whole blood expression</li>
<li>check how well the prediction works with GEUVADIS expression data</li>
<li>run association between predicted expression and a simulated phenotype</li>
<li>calculate association between expression levels and coronary artery disease risk using s-predixcan</li>
<li>fine-map the coronary artery disease gwas results using torus</li>
<li>calculate colocalization probability using fastenloc</li>
<li>run transcriptome-wide mendelian randomization in one locus of interestgi</li>
</ul>
</div>
<div id="initial-remarks" class="section level2">
<h2>Initial remarks</h2>
<ul>
<li>We ask you to actively participate in today’s hands on activities. <strong>Notice that we may ask you to share your screen for pedagogic purposes.</strong></li>
<li>If you have any concerns about this, please ask me or one of the TAs for assistance. We are here to help you learn.</li>
<li>As you run the analysis and programs, we ask you to <strong>find the tab with your name and respond the questions in <a href="https://docs.google.com/spreadsheets/d/1W_bFGB_ctLtUB3WTgAee4PohzKR-2lOKw9kTI0RVJr8/edit?usp=sharing">this document</a>.</strong> Find the tab with your name and fill out the questions as you go along.</li>
<li>You are welcome to check other people’s answers as guidelines but please make sure you write down your own answers.</li>
</ul>
</div>
<div id="preliminary-definitions" class="section level2">
<h2>Preliminary definitions</h2>
<ul class="task-list">
<li><input type="checkbox" disabled="" />
Go to the terminal tab on the RStudio server and run the following commands to update the analysis document to the most recent version.</li>
</ul>
<pre class="bash"><code>## Copy the lines below using Ctrl-C (or Cmd-C), paste on the terminal, and hit enter
PRE=&quot;/home/student/&quot;
cd $PRE/lab/
git pull </code></pre>
<ul class="task-list">
<li><p><input type="checkbox" disabled="" />
Under the Files tab (right bottom panel of RStudio) navigate to the folder Home/lab/analysis and open 2021_analysis_plan.Rmd (click on the file name)</p></li>
<li><p><input type="checkbox" disabled="" />
activate the the imlabtools environment, which will make sure all the necessary python modules are available to the software we will be running.</p></li>
</ul>
<pre class="bash"><code>## Copy the lines below using Ctrl-C (or Cmd-C), paste on the terminal, and hit enter
conda activate imlabtools</code></pre>
<p><strong>Reminder: the bash chunks need to be copy-pasted to the terminal, not performed within the chunk.</strong></p>
<ul class="task-list">
<li><input type="checkbox" disabled="" />
execute the following chunk to load the tidyverse package, a convenient set of R packages for analysis</li>
</ul>
<pre class="r"><code>suppressPackageStartupMessages(library(tidyverse))</code></pre>
<ul class="task-list">
<li><input type="checkbox" disabled="" />
To define some variables to access the data more easily within the R session, run the following r chunk</li>
</ul>
<pre class="r"><code>print(getwd())

lab=&quot;/home/student/lab&quot;
CODE=glue::glue(&quot;{lab}/code&quot;)
source(glue::glue(&quot;{CODE}/load_data_functions.R&quot;))
source(glue::glue(&quot;{CODE}/plotting_utils_functions.R&quot;))

PRE=&quot;/home/student/QGT-Columbia-HKI&quot;
MODEL=glue::glue(&quot;{PRE}/models&quot;)
DATA=glue::glue(&quot;{PRE}/data&quot;)
RESULTS=glue::glue(&quot;{PRE}/results&quot;)
METAXCAN=glue::glue(&quot;{PRE}/repos/MetaXcan-master/software&quot;)
FASTENLOC=glue::glue(&quot;{PRE}/repos/fastenloc-master&quot;)
TORUS=glue::glue(&quot;{PRE}/repos/torus-master&quot;)
TWMR=glue::glue(&quot;{PRE}/repos/TWMR-master&quot;)

# This is a reference table we&#39;ll use a lot throughout the lab. It contains information about the genes.
gencode_df = load_gencode_df()</code></pre>
<ul class="task-list">
<li><input type="checkbox" disabled="" />
define some variables to access the data more easily in the terminal. Remember we are running R code in the R console and command line code in the terminal.</li>
</ul>
<pre class="bash"><code>## Copy the lines below using Ctrl-C (or Cmd-C), paste on the terminal, and hit enter

export PRE=&quot;/home/student/QGT-Columbia-HKI&quot;
export LAB=&quot;/home/student/lab&quot;
export CODE=$LAB/code
export DATA=$PRE/data
export MODEL=$PRE/models
export RESULTS=$PRE/results
export METAXCAN=$PRE/repos/MetaXcan-master/software
export TWMR=$PRE/repos/TWMR-master</code></pre>
</div>
</div>
<div id="transcriptome-wide-association-methods" class="section level1">
<h1>Transcriptome-wide association methods</h1>
<p>Now we will perform a transcriptome-wide association analysis using the PrediXcan suite of tools. <img src="https://raw.githubusercontent.com/hakyimlab/QGT-Columbia-HKI/master/extras/figures/Association-Methods.png" alt="Transcriptome-wide association methods" /></p>
<p>We start by predicting the expression levels of genes using the genotype data and the prediction weights and then perform an association between the predicted expression and the phenotype (denoted trait in the figure below). <img src="https://raw.githubusercontent.com/hakyimlab/QGT-Columbia-HKI/master/extras/figures/PrediXcan-run.png" /></p>
<div id="predict-expression" class="section level2">
<h2>predict expression</h2>
<ul>
<li><p>We will predict expression of genes in whole blood using the Predict.py code in the METAXCAN folder.</p></li>
<li><p>Prediction models (weights) are located in the MODEL folder. Additional models for different tissues and transcriptome studies can be downloaded from <a href="http://predictdb.org">predictdb.org</a>.</p></li>
<li><p>Remember you need to copy and paste this code chunk into the terminal to run it. Also make sure you activated the imlabtools environment which has all the necessary python modules.</p></li>
<li><p>Make sure all the paths and file names are correct</p></li>
<li><p>This run should take about one minute.</p></li>
<li><p><input type="checkbox" disabled="" />
run the following code in the terminal.</p></li>
</ul>
<pre class="bash"><code>## Copy the lines below using Ctrl-C (or Cmd-C), paste on the terminal, and hit enter

printf &quot;Predict expression\n\n&quot;
python3 $METAXCAN/Predict.py \
--model_db_path $PRE/models/gtex_v8_en/en_Whole_Blood.db \
--vcf_genotypes $DATA/predixcan/genotype/filtered.vcf.gz \
--vcf_mode genotyped \
--variant_mapping $DATA/predixcan/gtex_v8_eur_filtered_maf0.01_monoallelic_variants.txt.gz id rsid \
--on_the_fly_mapping METADATA &quot;chr{}_{}_{}_{}_b38&quot; \
--prediction_output $RESULTS/predixcan/Whole_Blood__predict.txt \
--prediction_summary_output $RESULTS/predixcan/Whole_Blood__summary.txt \
--verbosity 9 \
--throw
</code></pre>
<ul class="task-list">
<li><input type="checkbox" disabled="" />
run following code in the console to get information on reported prediction performance.</li>
</ul>
<pre class="r"><code>prediction_fp = glue::glue(&quot;{RESULTS}/predixcan/Whole_Blood__predict.txt&quot;)

## Read the Predict.py output into a dataframe. This function reorganizes the data and adds gene names.
predicted_expression = load_predicted_expression(prediction_fp, gencode_df)

head(predicted_expression)

## read summary of prediction, number of SNPs per gene, cross validated prediction performance
prediction_summary = load_prediction_summary(glue::glue(&quot;{RESULTS}/predixcan/Whole_Blood__summary.txt&quot;), gencode_df)
## number of genes with a prediction model
dim(prediction_summary)
head(prediction_summary)

print(&quot;distribution of prediction performance r2&quot;)
summary(prediction_summary$pred_perf_r2)</code></pre>
</div>
<div id="assess-actual-prediction-performance-optional" class="section level2">
<h2>assess actual prediction performance (optional)</h2>
<pre class="r"><code>## download and read observed expression data from GEUVADIS 
## from https://uchicago.box.com/s/4y7xle5l0pnq9d1fwmthe2ewhogrnlrv

## Remove the version number from the gene_id&#39;s (ENSG000XXX.ver)
head(predicted_expression)

## merge predicted expression with observed expression data (by IID and gene)

## plot observes vs predicted expressioni for 
## ERAP1 (ENSG00000164307)
## PEX6 (ENSG00000124587)

## calculate spearman correlation for all genes

## what&#39;s the best performing gene?</code></pre>
</div>
<div id="run-association-with-a-simulated-phenotype" class="section level2">
<h2>run association with a simulated phenotype</h2>
<p><span class="math inline">\(Y = \sum_k T_k \beta_k + \epsilon\)</span></p>
<p>with random effects <span class="math inline">\(\beta_k \sim (1-\pi)\cdot \delta_0 + \pi\cdot N(0,1)\)</span></p>
<pre class="bash"><code>
export PHENO=&quot;sim.spike_n_slab_0.01_pve0.1&quot;

printf &quot;association\n\n&quot;
python3 $METAXCAN/PrediXcanAssociation.py \
--expression_file $RESULTS/predixcan/Whole_Blood__predict.txt \
--input_phenos_file $DATA/predixcan/phenotype/$PHENO.txt \
--input_phenos_column pheno \
--output $RESULTS/predixcan/$PHENO/Whole_Blood__association.txt \
--verbosity 9 \
--throw
</code></pre>
<p>More predicted phenotypes can be found in <code>$DATA/predixcan/phenotype/</code>. The naming of the phenotypes provides information about the genic architecture: the number after pve is the proportion of variance of Y explained by the genetic component of expression. The number after spike_n_slab represents the probability that a gene is causal <span class="math inline">\(\pi\)</span>(i.e. prob <span class="math inline">\(\beta \ne 0\)</span>)</p>
</div>
<div id="read-association-results" class="section level2">
<h2>read association results</h2>
<pre class="r"><code>## read association results
PHENO=&quot;sim.spike_n_slab_0.01_pve0.1&quot;

predixcan_association = load_predixcan_association(glue::glue(&quot;{RESULTS}/predixcan/{PHENO}/Whole_Blood__association.txt&quot;), gencode_df)

## take a look at the results
dim(predixcan_association)
predixcan_association %&gt;% arrange(pvalue) %&gt;% head
predixcan_association %&gt;% arrange(pvalue) %&gt;% ggplot(aes(pvalue)) + geom_histogram(bins=20)
## compare distribution against the null (uniform)
gg_qqplot(predixcan_association$pvalue, max_yval = 40)</code></pre>
</div>
<div id="compare-estimated-effects-with-true-effect-sizes" class="section level2">
<h2>compare estimated effects with true effect sizes</h2>
<pre class="r"><code>truebetas = load_truebetas(glue::glue(&quot;{DATA}/predixcan/phenotype/gene-effects/{PHENO}.txt&quot;), gencode_df)
betas = (predixcan_association %&gt;% 
               inner_join(truebetas,by=c(&quot;gene&quot;=&quot;gene_id&quot;)) %&gt;%
               select(c(&#39;estimated_beta&#39;=&#39;effect&#39;, 
                        &#39;true_beta&#39;=&#39;effect_size&#39;,
                        &#39;pvalue&#39;, 
                        &#39;gene_id&#39;=&#39;gene&#39;, 
                        &#39;gene_name&#39;=&#39;gene_name.x&#39;, 
                        &#39;region_id&#39;=&#39;region_id.x&#39;)))
betas %&gt;% arrange(pvalue) %&gt;% head
## do you see examples of potential LD contamination?
betas %&gt;% ggplot(aes(estimated_beta, true_beta))+geom_point()+geom_abline()</code></pre>
</div>
</div>
<div id="summary-predixcan" class="section level1">
<h1>Summary PrediXcan</h1>
<p>Now we will use the summary results from a GWAS of coronary artery disease to calculate the association between the genetic component of the expression of genes and coronary artery disease risk. We will use the SPrediXcan.py.</p>
<p><img src="https://raw.githubusercontent.com/hakyimlab/QGT-Columbia-HKI/master/extras/figures/gwas-PrediXcan-spredixcan.png" /></p>
<p>The GWAS results (harmonized and imputed) for coronary artery disease are available in $PRE/spredixcan/data/</p>
<div id="run-s-predixcan" class="section level2">
<h2>run s-predixcan</h2>
<pre class="bash"><code>python $METAXCAN/SPrediXcan.py \
--gwas_file  $DATA/spredixcan/imputed_CARDIoGRAM_C4D_CAD_ADDITIVE.txt.gz \
--snp_column panel_variant_id --effect_allele_column effect_allele --non_effect_allele_column non_effect_allele --zscore_column zscore \
--model_db_path $MODEL/gtex_v8_mashr/mashr_Whole_Blood.db \
--covariance $MODEL/gtex_v8_mashr/mashr_Whole_Blood.txt.gz \
--keep_non_rsid --additional_output --model_db_snp_key varID \
--throw \
--output_file $RESULTS/spredixcan/eqtl/CARDIoGRAM_C4D_CAD_ADDITIVE__PM__Whole_Blood.csv
</code></pre>
</div>
<div id="plot-and-interpret-s-predixcan-results" class="section level2">
<h2>plot and interpret s-predixcan results</h2>
<pre class="r"><code>spredixcan_association = load_spredixcan_association(glue::glue(&quot;{RESULTS}/spredixcan/eqtl/CARDIoGRAM_C4D_CAD_ADDITIVE__PM__Whole_Blood.csv&quot;), gencode_df)
dim(spredixcan_association)
spredixcan_association %&gt;% arrange(pvalue) %&gt;% head
spredixcan_association %&gt;% arrange(pvalue) %&gt;% ggplot(aes(pvalue)) + geom_histogram(bins=20)

gg_qqplot(spredixcan_association$pvalue)</code></pre>
<ul class="task-list">
<li><input type="checkbox" disabled="" />
SORT1, considered to be a causal gene for LDL cholesterol and as a consequence of coronary artery disease, is not found here. Why? (tissue)</li>
</ul>
</div>
<div id="exercise" class="section level2">
<h2>Exercise</h2>
<ul class="task-list">
<li><p><input type="checkbox" disabled="" />
run s-predixcan with liver model, do you find SORT1? Is it significant?</p></li>
<li><p><input type="checkbox" disabled="" />
compare zscores in liver and whole blood.</p></li>
</ul>
</div>
<div id="run-multixcan-optional" class="section level2">
<h2>run multixcan (optional)</h2>
<ul>
<li><p>multixcan aggregates information across multiple tissues to boost the power to detect association. It was developed movivated by the fact that eQTLs are shared across multiple tissues, i.e. many genetic variants that regulate expression are common across tissues.</p></li>
<li><p>before you run multixcan ensure you have run s-predixcan for all the tissues you want to multixcan. In this tutorial we have two tissues (liver and whole blood), ensure you have run s-predixcan with the two tissues before running multixcan.</p></li>
<li><p>One thing to note is to ensure similar naming pattern for the output files. This is to ensure the files are captured correctly when running multixcan’s filter.</p></li>
</ul>
<pre class="bash"><code>python $METAXCAN/SMulTiXcan.py \
--models_folder $MODEL/gtex_v8_mashr \
--models_name_pattern &quot;mashr_(.*).db&quot; \
--snp_covariance $MODEL/gtex_v8_expression_mashr_snp_smultixcan_covariance.txt.gz \
--metaxcan_folder $RESULTS/spredixcan/eqtl/ \
--metaxcan_filter &quot;CARDIoGRAM_C4D_CAD_ADDITIVE__PM__(.*).csv&quot; \
--metaxcan_file_name_parse_pattern &quot;(.*)__PM__(.*).csv&quot; \
--gwas_file $DATA/spredixcan/imputed_CARDIoGRAM_C4D_CAD_ADDITIVE.txt.gz \
--snp_column panel_variant_id --effect_allele_column effect_allele --non_effect_allele_column non_effect_allele --zscore_column zscore --keep_non_rsid --model_db_snp_key varID \
--cutoff_condition_number 30 \
--verbosity 7 \
--throw \
--output $RESULTS/smultixcan/eqtl/CARDIoGRAM_C4D_CAD_ADDITIVE_smultixcan.txt
</code></pre>
</div>
</div>
<div id="colocalization-methods" class="section level1">
<h1>Colocalization methods</h1>
<ul>
<li>Colocalization methods seek to estimate the probability that the complex trait and expression causal variants are the same. We favor methods that calculate the probability of causality for each trait (posterior inclusion probability), called fine-mapping methods. Here we use torus for fine-mapping and fastENLOC for colocalization.</li>
</ul>
<p><img src="https://raw.githubusercontent.com/hakyimlab/QGT-Columbia-HKI/master/extras/figures/colocalization-run.png" /></p>
<div id="gwas-summary-statistics-to-torus-format" class="section level2">
<h2>GWAS summary statistics to torus format</h2>
<ul>
<li><p>the following code will format GWAS summary statistics into a format that the fine-mapping method torus can understand.</p></li>
<li><p>we precalculated this for you so there is no need to run the following chunk</p></li>
</ul>
<pre class="bash"><code>## THERE IS NO NEED TO RUN THIS CHUNK
## We ran this formatting for you because it takes over 10 minutes.
python $CODE/gwas_to_torus_zscore.py \
-input_gwas $DATA/spredixcan/imputed_CARDIoGRAM_C4D_CAD_ADDITIVE.txt.gz \
-input_ld_regions $DATA/spredixcan/eur_ld_hg38.txt.gz \
-output_fp $DATA/fastenloc/CARDIoGRAM_C4D_CAD_ADDITIVE.zval.gz</code></pre>
</div>
<div id="fine-map-gwas-results" class="section level2">
<h2>fine-map GWAS results</h2>
<ul>
<li><p>We run torus due to time limitation but ideally we would like to run a method that allows multiple causal variants per locus, such as DAP-G or SusieR.</p></li>
<li><p>torus has been precompiled and placed within the PATH</p></li>
</ul>
<pre class="bash"><code>export TORUSOFT=torus

$TORUSOFT -d $PRE/data/fastenloc/CARDIoGRAM_C4D_CAD_ADDITIVE.zval.gz --load_zval -dump_pip $PRE/data/fastenloc/CARDIoGRAM_C4D_CAD_ADDITIVE.gwas.pip
cd $PRE/data/fastenloc
gzip CARDIoGRAM_C4D_CAD_ADDITIVE.gwas.pip
cd $PRE </code></pre>
<p>We can take a quick look at the z-values and finemapping PIPs (posterior probability of causality):</p>
<pre class="bash"><code>cd $PRE/data/fastenloc
zless CARDIoGRAM_C4D_CAD_ADDITIVE.zval.gz
zless CARDIoGRAM_C4D_CAD_ADDITIVE.gwas.pip.gz</code></pre>
</div>
<div id="calculate-colocalization-with-fastenloc" class="section level2">
<h2>calculate colocalization with fastENLOC</h2>
<pre class="bash"><code>## you can take a look at the tutorial https://github.com/xqwen/fastenloc/tree/master/tutorial

export eqtl_annotation_gzipped=$PRE/data/fastenloc/FASTENLOC-gtex_v8.eqtl_annot.vcf.gz
export gwas_data_gzipped=$PRE/data/fastenloc/CARDIoGRAM_C4D_CAD_ADDITIVE.gwas.pip.gz
export TISSUE=Whole_Blood
export FASTENLOCSOFT=fastenloc
##export FASTENLOCSOFT=/Users/owenmelia/projects/finemapping_bin/src/fastenloc/src/fastenloc

mkdir $RESULTS/fastenloc/
cd $RESULTS/fastenloc/
$FASTENLOCSOFT -eqtl $eqtl_annotation_gzipped -gwas $gwas_data_gzipped -t $TISSUE 

#[-total_variants total_snp] [-thread n] [-prefix prefix_name] [-s shrinkage]
</code></pre>
</div>
<div id="analyze-results" class="section level2">
<h2>analyze results</h2>
<pre class="r"><code>## optional - compare with s-predixcan results

fastenloc_results = load_fastenloc_coloc_result(glue::glue(&quot;{RESULTS}/fastenloc/enloc.sig.out&quot;))
spredixcan_and_fastenloc = inner_join(spredixcan_association, fastenloc_results, by=c(&#39;gene&#39;=&#39;Signal&#39;))
ggplot(spredixcan_and_fastenloc, aes(RCP, -log10(pvalue))) + geom_point()

## which genes are both colocalized (rcp&gt;0.10) and significantly associated (pvalue&lt;0.05/number of tests)</code></pre>
<hr />
</div>
</div>
<div id="mendelian-randomization-methods" class="section level1">
<h1>Mendelian randomization methods</h1>
<div id="run-twmr-for-a-locus" class="section level2">
<h2>run TWMR (for a locus)</h2>
<div class="figure">
<img src="https://raw.githubusercontent.com/hakyimlab/QGT-Columbia-HKI/master/extras/figures/TWMR.png" alt="" />
<p class="caption">TWMR</p>
</div>
<pre class="r"><code># Load the &#39;analyseMR&#39; function
source(glue::glue(&quot;{CODE}/TWMR_script.R&quot;))

# Collect the list of genes available to run
gene_lst &lt;- list.files(TWMR)
gene_lst &lt;- gene_lst[str_detect(gene_lst, &quot;ENS.*&quot;)]
gene_lst &lt;- (gsub(&quot;\\..*&quot;, &quot;&quot;, gene_lst) %&gt;% unique)

# Set the gene and run. The function writes output to a file.
for (gene in gene_lst) {
  analyseMR(gene, TWMR)
}</code></pre>
<pre class="r"><code>twmr_results &lt;- load_twmr_results(TWMR, gencode_df)</code></pre>
</div>
</div>
<div id="optional-items" class="section level1">
<h1>Optional items</h1>
<div id="setting-up-your-own-system" class="section level2">
<h2>Setting up your own system</h2>
<p>Linux is the operating system of choice to run bioinformatics software. You will need either a computer running linux or or mac os, which has a linux-like environment.</p>
<ul class="task-list">
<li><input type="checkbox" disabled="" />
install anaconda/miniconda</li>
<li><input type="checkbox" disabled="" />
define imlabtools conda environment <a href="https://github.com/hakyimlab/MetaXcan/blob/master/README.md#example-conda-environment-setup">how to here</a>, which will install all the python modules needed for this analysis session</li>
<li><input type="checkbox" disabled="" />
download data and software <a href="https://uchicago.box.com/s/zhapf2zfxcpj7thvq4sjnqale3emleum">from Box</a>. This will have copies of all the software repositories and the models</li>
<li><input type="checkbox" disabled="" />
download software<ul>
<li>download metaxcan repo</li>
<li>download torus repo</li>
<li>download fastenloc repo</li>
<li>download TMWR repo</li>
</ul></li>
<li><input type="checkbox" disabled="" />
download prediction models from predictdb.org</li>
<li><input type="checkbox" disabled="" />
install R/RStudio/tidyverse package</li>
<li><input type="checkbox" disabled="" />
(optional) install workflowr package in R</li>
<li><input type="checkbox" disabled="" />
git clone <a href="https://github.com/hakyimlab/QGT-Columbia-HKI.git" class="uri">https://github.com/hakyimlab/QGT-Columbia-HKI.git</a></li>
<li><input type="checkbox" disabled="" />
start Rstudio (if you installed workflowr, you can just open the QGT-Columbia-HKI.Rproj)</li>
</ul>
</div>
<div id="how-to-create-rstudio-server" class="section level2">
<h2>How to create RStudio server</h2>
<p>Check instructions <a href="https://lab-notes.hakyimlab.org/post/2020/06/16/introstatgen-r-studio-servers-using-google-cloud/">here</a></p>
</div>
<div id="slides" class="section level2">
<h2>Slides</h2>
<p>Download slides <a href="https://github.com/hakyimlab/QGT-Columbia-HKI-repo/blob/master/extras/QGT-Columbia-2021-HKI.pdf">here</a></p>
</div>
</div>
<div id="acknowledgements" class="section level1">
<h1>Acknowledgements</h1>
<p>Contributions by Owen Melia, Yanyu Liang, and Tyson Miller</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.3 (2020-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5       rstudioapi_0.11  whisker_0.4      knitr_1.30      
 [5] magrittr_2.0.1   R6_2.5.0         rlang_0.4.10     stringr_1.4.0   
 [9] tools_3.6.3      xfun_0.18        git2r_0.27.1     htmltools_0.5.0 
[13] ellipsis_0.3.1   rprojroot_1.3-2  yaml_2.2.1       digest_0.6.27   
[17] tibble_3.0.4     lifecycle_0.2.0  crayon_1.3.4     later_1.1.0.1   
[21] vctrs_0.3.5      promises_1.1.1   fs_1.5.0         glue_1.4.2      
[25] evaluate_0.14    rmarkdown_2.5    stringi_1.5.3    compiler_3.6.3  
[29] pillar_1.4.6     backports_1.1.10 httpuv_1.5.4     pkgconfig_2.0.3 </code></pre>
</div>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>note that these servers are different from the Rstudio cloud you used yesterday<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
